### DISCUSSION TITLE
What is the replacement for `X509Certificate.CreateFromSignedFile`?

### DISCUSSION THREAD

**Discussion by bgrainger** [dotnet/runtime#108740] [Q&A] [answered]
The following code works under .NET 8 (on Windows) and prints `CN=Microsoft Windows Production PCA 2011, O=Microsoft Corporation, L=Redmond, S=Washington, C=US`:

```csharp
var cert = X509Certificate.CreateFromSignedFile(@"C:\Windows\explorer.exe");
Console.WriteLine(cert.Issuer);
```

Under .NET 9, compiling that code gives the warning:

> warning SYSLIB0057: 'X509Certificate.CreateFromSignedFile(string)' is obsolete: 'Loading certificate data through the constructor or Import is obsolete. Use X509CertificateLoader instead to load certificates.' (https://aka.ms/dotnet-warnings/SYSLIB0057)

Firstly, I don't completely understand the warning, because I'm not using a constructor or `Import`. I'm using a specific factory method already.

I've read https://github.com/dotnet/runtime/issues/91763, https://github.com/dotnet/docs/issues/41662, and https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0057#workaround but I haven't seen a replacement given for this specific method.

`X509Certificate2.GetCertContentType(@"C:\Windows\explorer.exe")` returns `Authenticode`.

What is the right, non-obsolete way to check an Authenticode signature in .NET 9?

**Comment by bgrainger** [comment]
Is the answer to install [Microsoft.Security.Extensions](https://www.nuget.org/packages/Microsoft.Security.Extensions) and do the following?

```csharp
using var stream = File.OpenRead(@"C:\Windows\explorer.exe");
var signatureInfo = FileSignatureInfo.GetFromFileStream(stream);
Console.WriteLine(signatureInfo.SigningCertificate.Issuer);
```

That is, has this functionality moved out of core .NET in .NET 9 and now needs to be supplied by a NuGet package?

**Reply by bgrainger** [reply]
If this is the recommended answer, then the `[Obsolete]` message and the docs linked in the OP could be updated to make this much clearer.

**Reply by bartonjs** [reply, marked as answer]
That package is not the recommended answer, because I don't know where it comes from.  It might be fine, I just couldn't establish confidence in its maintenance with 2 minutes of looking.

The recommended answer, if you're wanting to extract the Authenticode signer, is to combine GetCertContentType and the ctor and suppress the warning.  The reason CreateFromSignedFile is included in the obsoletion is that it doesn't only load Authenticode, it's always been a poorly named vanity wrapper around the "do everything" constructor (it's just implemented as `return new X509Certificate2(path)`)

```C#
private static X509Certificate2 LoadAuthenticode(string path)
{
    if (X509Certificate2.GetCertContentType(path) == X509ContentType.Authenticode)
    {
#pragma warning disable SYSLIB0057
        return new X509Certificate2(path);
#pragma warning restore SYSLIB0057
    }

    throw new CryptographicException();
}
```

**Reply by whiskhub** [reply]
But obsoleted methods will get removed at some point, right? 
Shouldn’t a X509CertificateLoader.LoadAuthenticode() method be added to the framework?

**Reply by bartonjs** [reply]
> But obsoleted methods will get removed at some point, right?

Not generally.  We might hollow out the method (make it throw PlatformNotSupportedException), but generally not "remove".  And until we drop support for .NET Framework assemblies using "faith call", or libraries using .NET Standard, we have to expect that people will still be calling these methods as "I'm on an older framework version"... so we won't be hollowing them out.

The reason the methods are obsolete is that more than 95% of callers have a better method to call now.

> Shouldn’t a X509CertificateLoader.LoadAuthenticode() method be added to the framework?

Extracting the Authenticode signer cert has turned out to be surprisingly popular, for a feature I didn't know was part of the framework until I owned the code and had to maintain it.

It was part of the original proposal (https://github.com/dotnet/runtime/issues/91763#issue-1886608804, top post, last edit before October 10th 2023):

```C#
 // Whatever new X509Certificate2("some.exe") does.
// I think it extracts the SignedCms and then is the same as the next set.
[SupportedOS(Windows)]
public static partial X509Certificate2 LoadAuthenticodeSigner(byte[] data);
[SupportedOS(Windows)]
public static partial X509Certificate2 LoadAuthenticodeSigner(ReadOnlySpan<byte> data);
[SupportedOS(Windows)]
public static partial X509Certificate2 LoadAuthenticodeSigner(string path);
```

When it underwent API Review, we cut all of the things that only worked on one OS (which was always Windows).  But perhaps we'll have to reconsider this one for .NET 11.

**Reply by chrarnoldus** [reply]
> until we drop support for .NET Framework assemblies using "faith call", or libraries using .NET Standard

Is this expected to happen eventually or is it just a hypothetical?

**Reply by bartonjs** [reply]
There is no active plan that I know of to drop support for either of those things; I was just saying that in this case it would be a pre-req.  So... "hypothetical".

**Reply by ghost** [reply]
> But obsoleted methods will get removed at some point, right? Shouldn’t a X509CertificateLoader.LoadAuthenticode() method be added to the framework?

I hope this will be implemented. I use Authenticode too for file integrity checks.

**Reply by rbdavison** [reply]
None of the previous answers or responses did what I needed. I'd really like to get all certificates from signed file to be able to perform checks against them.
The solution for my case was the following code
```
using System.Reflection.PortableExecutable;
using System.Security.Cryptography;
using System.Security.Cryptography.Pkcs;
using System.Security.Cryptography.X509Certificates;

namespace MyNamespace
{
    internal static class GetCerts
    {
        /// <summary>
        /// Gets all the signing certificates as a collection.
        /// </summary>
        /// <param name="pePath"></param>
        /// <returns></returns>
        internal static X509Certificate2Collection GetSigningCertificates(string pePath)
        {
            X509Certificate2Collection certificateCollection = new X509Certificate2Collection();
            try
            {
                using (FileStream fs = new FileStream(pePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
                using (PEReader peReader = new PEReader(fs))
                {
                    // Display PE header information
                    if (peReader.PEHeaders.PEHeader is PEHeader header)
                    {
                        DirectoryEntry certificateTableDirectoryEntry = header.CertificateTableDirectory;
                        if (certificateTableDirectoryEntry.Size > 0)
                        {
                            fs.Position = certificateTableDirectoryEntry.RelativeVirtualAddress;
                            using (BinaryReader reader = new BinaryReader(fs))
                            {
                                int bytesRead = 0;
                                while (bytesRead < certificateTableDirectoryEntry.Size)
                                {
                                    // WIN_CERTIFICATE structure
                                    uint dwLength = reader.ReadUInt32(); // Length of total entry
                                    uint wRevision = reader.ReadUInt16(); // wRevision
                                    uint wCertificateType = reader.ReadUInt16(); // wCertificateType

                                    // Length includes header (8 bytes)
                                    int certPayloadLength = (int)dwLength - 8;
                                    byte[] certBytes = reader.ReadBytes(certPayloadLength);

                                    // Parse the PKCS#7 SignedData
                                    try
                                    {
                                        // Signer certs are usually in the PKCS#7 wrapper
                                        SignedCms cms = new SignedCms();
                                        cms.Decode(certBytes);
                                        foreach (SignerInfo signerInfo in cms.SignerInfos)
                                        {
                                            SubjectIdentifier signerIdentifier = signerInfo.SignerIdentifier;
                                            if (signerInfo.Certificate is X509Certificate2 cert)
                                            {
                                                _ = certificateCollection.Add(cert);
                                            }
                                            foreach (SignerInfo counterSigner in signerInfo.CounterSignerInfos)
                                            {
                                                SubjectIdentifier counterSignerIdentifier = counterSigner.SignerIdentifier;
                                                if (counterSigner.Certificate is X509Certificate2 counterCert)
                                                {
                                                    _ = certificateCollection.Add(counterCert);
                                                }
                                            }
                                            foreach (CryptographicAttributeObject unsignedAttributes in signerInfo.UnsignedAttributes)
                                            {
                                                foreach (AsnEncodedData value in unsignedAttributes.Values)
                                                {
                                                    try
                                                    {
                                                        SignedCms signedCms = new SignedCms();
                                                        signedCms.Decode(value.RawData);
                                                        foreach (X509Certificate2 extraCert in signedCms.Certificates)
                                                        {
                                                            _ = certificateCollection.Add(extraCert);
                                                        }
                                                    }
                                                    catch (Exception ex)
                                                    {
                                                        Console.WriteLine(ex);
                                                    }
                                                }
                                            }
                                        }
                                    }
                                    catch (Exception ex)
                                    {
                                        Console.WriteLine($"Error parsing cert: {ex.Message}");
                                    }

                                    // Align to 8-byte boundary (PE requirement)
                                    int padding = (8 - (int)(dwLength % 8)) % 8;
                                    _ = reader.ReadBytes(padding);
                                    bytesRead += (int)dwLength + padding;
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex);
            }
            return certificateCollection;
        }
    }
}
```
